### DISCUSSION TITLE
Dlib GUI

### DISCUSSION THREAD

**Discussion by pfeatherstone** [davisking/dlib#2745] [General]
Are there any plans in the future to work on the GUI stuff?

I typically use web GUIs but I'm starting to realise that a lot of the web tooling is designed to work best with node.js and/or large frameworks like React, Vue or Angular. Gone are the days where web libraries could be simply included in your HTML header and used JQuery at most. So I've started to look at desktop GUI libraries like GTK, imgui, QT and dlib. 

So I was wondering what the state of dlib GUI was in? 

I'm interested in scientific plotting, so line graphs, scatter graphs and heat maps for spectrograms. Those would need a PR and possibly use dlib's canvas API. But is the X11 backend or Windows backend suitable for plotting large amounts of data? Or would dlib require another set of PRs to support OpenGL, Vulcan or whatever for rendering large amounts of data? 

At the moment, using another library would be more suitable. But I'm interested to know if someone had plans to pimp up dlib's GUI.

**Comment by davisking** [comment]
It's great.  I use it all the time.  Works smoothly through X11 forwarding through ssh too.  I've already made plenty of 2D and 3D scatter plots with dlib's GUI.  It is super fast for that.  All it is is a memory buffer that's like "ok, write to these pixels however you want".  It isn't going to be faster than that, since your user code of any other API is still going to have to run and loop over all your points or whatever it is you are dealing with anyway.  If you wanted to draw a bunch of 3D triangles and do shading and stuff like that then OpenGL would be faster obviously.  But the math for a scatter plot is not meaningfully different from a memory copy.

Even drawing a bunch of lines is real fast.  Heatmaps are easy too.  Like look at http://dlib.net/imaging.html#heatmap.  Or go make http://dlib.net/3d_point_cloud_ex.cpp.html plot millions of points.  Renders fine.  Even on the other end of X11 forwarding, which would just break if you tried it with opengl.

A lot of these things just boil down to making and image somehow.  It's not the GUI libraries business how that happens.  Then you put it on the screen with `image_window`.

**Reply by pfeatherstone** [reply]
Interesting. How about styles for background colours, line widths, fonts, things like that to jazz up the UI. Is that on a roadmap or something?

**Reply by davisking** [reply]
I'm not, I'm happy with what's there now.  You can add whatever though.

**Reply by pfeatherstone** [reply]
QT supports stylesheets, similar to CSS. Something like that in dlib would be awesome to make GUIs look a bit jazzier. No idea how much effort that is.

**Reply by davisking** [reply]
Yeah, would be neat but that's a ton of work. Dlib's GUI intentionally doesn't depend on anything.  It does all the drawing with code that has no dependencies, so it's super portable.  The positioning of widgets is also just by pixel location, which makes a ton of things easy for the user, but makes having stylesheets hard since you lose the ability to automatically reflow layouts.

You can make different kinds of styles though like http://dlib.net/dlib/gui_widgets/style_abstract.h.html.  That's easy to do.

**Reply by pfeatherstone** [reply]
Yeah that's great. honestly, I don't care about styling too much, so long as it doesn't look like it was made in 1980. I care more about something that's fast to develop, functional and fast. To be honest, dlib looks like a good candidate. I would just need to add some plotting widgets for line curves, heatmaps (spectrograms), scatter plots, all with some zoom options, panning, maybe a grid layout.

**Reply by pfeatherstone** [reply]
Would you consider open sourcing some example GUIs you've made with dlib, in addition to imglab, including some where you've customized certain bits? Maybe even the plotting widgets you've mentioned.

**Reply by davisking** [reply]
I don't have any cool GUI stuff that isn't in dlib.  I mean, I throw stuff together but but nothing fancy. Like the heatmap stuff I mentioned is in dlib now.  Mostly whenever I do any 2d plotting I just toss it into matplotlib or GNU octave.
